src/algorithms/graph/detect-cycle/detectUndirectedCycleUsingDisjointSet.js

```diff
@@ -6,26 +6,19 @@ import DisjointSet from '../../../data-structures/disjoint-set/DisjointSet';
  * @param {Graph} graph
  */
 export default function detectUndirectedCycleUsingDisjointSet(graph) {
-  // Create initial singleton disjoint sets for each graph vertex.
-  /** @param {GraphVertex} graphVertex */
-  const keyExtractor = graphVertex => graphVertex.getKey();
-  const disjointSet = new DisjointSet(keyExtractor);
-  graph.getAllVertices().forEach(graphVertex => disjointSet.makeSet(graphVertex));
+  const vertices = graph.getAllVertices();
+  const disjointSet = new DisjointSet(vertices.length);
 
-  // Go trough all graph edges one by one and check if edge vertices are from the
-  // different sets. In this case joint those sets together. Do this until you find
-  // an edge where to edge vertices are already in one set. This means that current
-  // edge will create a cycle.
-  let cycleFound = false;
-  /** @param {GraphEdge} graphEdge */
-  graph.getAllEdges().forEach((graphEdge) => {
-    if (disjointSet.inSameSet(graphEdge.startVertex, graphEdge.endVertex)) {
-      // Cycle found.
-      cycleFound = true;
-    } else {
-      disjointSet.union(graphEdge.startVertex, graphEdge.endVertex);
+  for (let i = 0; i < vertices.length; i += 1) {
```

Comment:
I still prefer object-oriented manner of dealing with set elements instead of index-based. Just for learning and code reading purposes.

Take a look ta these two examples:
```javascript
graph.getAllEdges().forEach((graphEdge) => {
  if (disjointSet.inSameSet(graphEdge.startVertex, graphEdge.endVertex)) {
    ...
  }
}
```

and 

```
for (let i = 0; i < vertices.length; i += 1) {
  for (let j = (i + 1); j < vertices.length; j += 1) {
    if (vertices[i].hasNeighbor(vertices[j])) {
      if (disjointSet.find(i) === disjointSet.find(j)) {
        ...
      }
    }
  }
}
```

It feels like the first one is more readable.

Reply:
I do not think that this is a reasonable comparison.  Please let me know if I am wrong, but it very much seemed to me that the implementation of disjoint set was extremely influenced by the current implementation of graph.  I didn't think this was appropriate, as they are two separate data structures whose various implementations should be clear and distinct.  It is thus no surprise that the first pass at integration looks clunky.

The crux of the integration issue here is that there is no elegant way to map vertices to indices in this graph implementation.  However, there are a variety of other implementations of graphs that provide such a transformation very naturally.  Even if this is not the case, the only roadblock is the creation of a vertex to index mapping, which we could easily showcase.  Thus, I don't think that a repository of data structures & algorithms like this one should try to lock people in to the current implementation.  Part of being a software engineer is the ability to glue together the appropriate components in a pleasing way.

In short, I think that this post optimizes for simplicity in the wrong place; I think it is the data structures that should be straightforward.  I strongly believe that, in the long view, the composition of simple data structures will lead to more beautiful implementations of these algorithms.